# GBRAP
This repository contains the codes used in the GBRAP (Genome-Based Retrieval and Analysis Parser) project. 

# Overview
A comprehensive software tool to analyse GenBank files and an online database, housing an extensive collection of carefully curated, high-quality genome statistics generated by the software for all the organisms (latest assembly; reference or representative genome) available in the RefSeq database of NCBI. The users can directly search or select from pre-categorized groups, the organisms of their choice and access data tables containing useful genomic information (eg. Base counts, GC content, Shannon Entropy, Chargaff Score) separately calculated for different genomic elements (eg. CDS, Introns, ncRNA) adding up to a collection of more than 200 genomic statistics. The data are separately displayed (if applicable) for each chromosomal, mitochondrial or chloroplast sequence. All the data can be visualised on the database itself or downloaded as CSV or Excel files. The GBRAP database is free to access without any registration and is publicly available at <add GBRAP database link>

# Installation and Usage
GBRAP is an open-source tool and is freely available at <insert github link>. Once the user has the gbff file needed to be analysed GBRAP can be used in the command line/ terminal by running the command “. /GBRAP_Args.py in input_file_name.gbff out output_file_name.csv”. GBRAP includes all the needed files for the execution of the complete analysis, and no additional installations are necessary. 

GBRAP returns in output a comma-separated file (CSV) that could be imported into excel. Each file contains:\n
Locus ID: Sequence ID\n
Definition: Sequence description\n
bp_chromo_A: Number of 'A' bases in the whole chromosome/genome\n
bp_chromo_T: Number of 'T' bases in the whole chromosome/genome\n
•	bp_chromo_C: Number of 'C' bases in the whole chromosome/genome
•	bp_chromo_G: Number of 'G' bases in the whole chromosome/genome
•	bp_chromo_N: Number of 'N' bases in the whole chromosome/genome
•	bp_chromo_total: Total number of bases in the whole chromosome/genome
•	fr_chromo_A: Frequency of 'A' bases among total the whole chromosome/genome bases
•	fr_chromo_T: Frequency of 'T' bases among total the whole chromosome/genome bases
•	fr_chromo_C: Frequency of 'C' bases among total the whole chromosome/genome bases
•	fr_chromo_G: Frequency of 'G' bases among total the whole chromosome/genome bases
•	fr_chromo_N: Frequency of 'N' bases among total the whole chromosome/genome bases
•	GC_chromo: Percentage of 'G' and 'C' bases in the whole chromosome/genome
•	chromo_topological_entropy: Topological entropy calculated from the whole chromosome/genome
•	chromo_chargaff_pf: Chargaff's second parity rule score (method1) for the whole chromosome/genome
•	chromo_chargaff_ct: Chargaff's second parity rule score (method2) for the whole chromosome/genome
•	chromo_shannon: Shannon entropy score for the whole chromosome/genome
•	n_gene_pos: Number of genes in the positive strand
•	n_gene_neg: Number of genes in the negative strand
•	n_gene_tot: Total number of genes in the chromosome
•	bp_gene_A: Number of 'A' bases in total genes
•	bp_gene_T: Number of 'T' bases in total genes
•	bp_gene_C: Number of 'C' bases in total genes
•	bp_gene_G: Number of 'G' bases in total genes
•	bp_gene_N: Number of 'N' bases in total genes
•	bp_gene_total: Total number of bases in total genes
•	fr_gene_A: Frequency of 'A' bases among total gene bases
•	fr_gene_T: Frequency of 'T' bases among total gene bases
•	fr_gene_C: Frequency of 'C' bases among total gene bases
•	fr_gene_G: Frequency of 'G' bases among total gene bases
•	fr_gene_N: Frequency of 'N' bases among total gene bases
•	GC_gene: Percentage of 'G' and 'C' bases in total genes
•	gene_topological_entropy: Topological entropy calculated from total gene sequences
•	gene_chargaff_pf: Chargaff's second parity rule score (method1) for total genes
•	gene_chargaff_ct: Chargaff's second parity rule score (method2)for total genes
•	gene_shannon: Shannon entropy score for total genes
•	bp_gene_overlap_total: Number of bases overlapping between genes on two strands
•	n_cds_pos: Number of CDS in the positive strand
•	n_cds_neg: Number of CDS in the negative strand
•	n_cds_tot: Total number of CDS in the chromosome
•	bp_cds_A: Number of 'A' bases in total CDS (coding sequences)
•	bp_cds_T: Number of 'T' bases in total CDS
•	bp_cds_C: Number of 'C' bases in total CDS
•	bp_cds_G: Number of 'G' bases in total CDS
•	bp_cds_N: Number of 'N' bases in total CDS
•	bp_cds_total: Total bases in total CDS
•	fr_cds_A: Frequency of 'A' bases among total CDS bases
•	fr_cds_T: Frequency of 'T' bases among total CDS bases
•	fr_cds_C: Frequency of 'C' bases among total CDS bases
•	fr_cds_G: Frequency of 'G' bases among total CDS bases
•	fr_cds_N: Frequency of 'N' bases among total CDS bases
•	GC_cds: Percentage of 'G' and 'C' bases in total CDS
•	cds_topological_entropy: Topological entropy calculated from total CDS sequences
•	cds_chargaff_pf: Chargaff's second parity rule score (method1) for total CDS
•	cds_chargaff_ct: Chargaff's second parity rule score (method2) for total CDS
•	cds_shannon: Shannon entropy score for total CDS
•	bp_cds_overlap_total: Number of bases overlapping between CDS on two strands
•	bp_cds_intron_A: Number of 'A' bases in total introns between CDS
•	bp_cds_intron_T: Number of 'T' bases in total introns between CDS
•	bp_cds_intron_C: Number of 'C' bases in total introns between CDS
•	bp_cds_intron_G: Number of 'G' bases in total introns between CDS
•	bp_cds_intron_N: Number of 'N' bases in total introns between CDS
•	bp_cds_intron_total: Total bases in total introns between CDS
•	fr_cds_intronn_A: Frequency of 'A' bases in total introns between CDS
•	fr_cds_intron_T: Frequency of 'T' bases in total introns between CDS
•	fr_cds_intron_C: Frequency of 'C' bases in total introns between CDS
•	fr_cds_intron_G: Frequency of 'G' bases in total introns between CDS
•	fr_cds_intron_N: Frequency of 'N' bases in total introns between CDS
•	GC_cds_intron: Percentage of 'G' and 'C' bases in total introns between CDS
•	cds_intron_topological_entropy: Topological entropy calculated from total intron sequences between CDS
•	cds_intron_chargaff_pf: Chargaff's second parity rule score (method1) for total introns between CDS
•	cds_intron_chargaff_ct: Chargaff's second parity rule score (method2) for total introns between CDS
•	cds_intron_shannon: Shannon entropy score for total introns between CDS
•	bp_cds_intron_overlap_total: Number of bases overlapping between introns between CDS on two strands
•	n_ncRNA_pos: Number of ncRNA in the positive strand
•	n_ncRNA_neg: Number of ncRNA in the negative strand
•	n_ncRNA_tot: Total number of ncRNA in the chromosome
•	bp_ncRNA_A: Number of 'A' bases in total ncRNA
•	bp_ncRNA_T: Number of 'T' bases in total ncRNA
•	bp_ncRNA_C: Number of 'C' bases in total ncRNA
•	bp_ncRNA_G: Number of 'G' bases in total ncRNA
•	bp_ncRNA_N: Number of 'N' bases in total ncRNA
•	bp_ncRNA_total: Total bases in total ncRNA
•	fr_ncRNA_A: Frequency of 'A' bases among total ncRNA bases
•	fr_ncRNA_T: Frequency of 'T' bases among total ncRNA bases
•	fr_ncRNA_C: Frequency of 'C' bases among total ncRNA bases
•	fr_ncRNA_G: Frequency of 'G' bases among total ncRNA bases
•	fr_ncRNA_N: Frequency of 'N' bases among total ncRNA bases
•	GC_ncRNA: Percentage of 'G' and 'C' bases in total ncRNA
•	ncRNA_topological_entropy: Topological entropy calculated from total ncRNA sequences
•	ncRNA_chargaff_pf: Chargaff's second parity rule score (method1) for total ncRNA
•	ncRNA_chargaff_ct: Chargaff's second parity rule score (method2) for total ncRNA
•	ncRNA_shannon: Shannon entropy score for total ncRNA
•	bp_ncRNA_overlap_total: Number of bases overlapping between ncRNA on two strands
•	bp_ncintron_A: Number of 'A' bases in total introns between ncRNA
•	bp_ncintron_T: Number of 'T' bases in total introns between ncRNA
•	bp_ncintron_C: Number of 'C' bases in total introns between ncRNA
•	bp_ncintron_G: Number of 'G' bases in total introns between ncRNA
•	bp_ncintron_N: Number of 'N' bases in total introns between ncRNA
•	bp_ncintron_total: Total bases in total introns between ncRNA
•	fr_ncintron_A: Frequency of 'A' bases in total introns between ncRNA
•	fr_ncintron_T: Frequency of 'T' bases in total introns between ncRNA
•	fr_ncintron_C: Frequency of 'C' bases in total introns between ncRNA
•	fr_ncintron_G: Frequency of 'G' bases in total introns between ncRNA
•	fr_ncintron_N: Frequency of 'N' bases in total introns between ncRNA
•	GC_ncintron: Percentage of 'G' and 'C' bases in total introns between ncRNA
•	ncintron_topological_entropy: Topological entropy calculated from total intron sequences between ncRNA
•	ncintron_chargaff_pf: Chargaff's second parity rule score (method1) for total introns between ncRNA
•	ncintron_chargaff_ct: Chargaff's second parity rule score (method2) for total introns between ncRNA
•	ncintron_shannon: Shannon entropy score for total introns between ncRNA
•	bp_ncintron_overlap_total: Number of bases overlapping between introns between ncRNA on two strands
•	n_tRNA_pos: Number of tRNA in the positive strand
•	n_tRNA_neg: Number of tRNA in the negative strand
•	n_tRNA_tot: Total number of tRNA in the chromosome
•	bp_tRNA_A: Number of 'A' bases in total tRNA
•	bp_tRNA_T: Number of 'T' bases in total tRNA
•	bp_tRNA_C: Number of 'C' bases in total tRNA
•	bp_tRNA_G: Number of 'G' bases in total tRNA
•	bp_tRNA_N: Number of 'N' bases in total tRNA
•	bp_tRNA_total: Total bases in total tRNA
•	fr_tRNA_A: Frequency of 'A' bases among total tRNA bases
•	fr_tRNA_T: Frequency of 'T' bases among total tRNA bases
•	fr_tRNA_C: Frequency of 'C' bases among total tRNA bases
•	fr_tRNA_G: Frequency of 'G' bases among total tRNA bases
•	fr_tRNA_N: Frequency of 'N' bases among total tRNA bases
•	GC_tRNA: Percentage of 'G' and 'C' bases in total tRNA
•	tRNA_topological_entropy: Topological entropy calculated from total tRNA sequences
•	tRNA_chargaff_pf: Chargaff's second parity rule score (method1) for total tRNA
•	tRNA_chargaff_ct: Chargaff's second parity rule score (method2) for total tRNA
•	tRNA_shannon: Shannon entropy score for total tRNA
•	bp_tRNA_overlap_total: Number of bases overlapping between tRNA on two strands
•	n_rRNA_pos: Number of rRNA in the positive strand
•	n_rRNA_neg: Number of rRNA in the negative strand
•	n_rRNA_tot: Total number of rRNA in the chromosome
•	bp_rRNA_A: Number of 'A' bases in total rRNA
•	bp_rRNA_T: Number of 'T' bases in total rRNA
•	bp_rRNA_C: Number of 'C' bases in total rRNA
•	bp_rRNA_G: Number of 'G' bases in total rRNA
•	bp_rRNA_N: Number of 'N' bases in total rRNA
•	bp_rRNA_total: Total bases in total rRNA
•	fr_rRNA_A: Frequency of 'A' bases among total rRNA bases
•	fr_rRNA_T: Frequency of 'T' bases among total rRNA bases
•	fr_rRNA_C: Frequency of 'C' bases among total rRNA bases
•	fr_rRNA_G: Frequency of 'G' bases among total rRNA bases
•	fr_rRNA_N: Frequency of 'N' bases among total rRNA bases
•	GC_rRNA: Percentage of 'G' and 'C' bases in total rRNA
•	rRNA_topological_entropy: Topological entropy calculated from total rRNA sequences
•	rRNA_chargaff_pf: Chargaff's second parity rule score (method1) for total rRNA
•	rRNA_chargaff_ct: Chargaff's second parity rule score (method2) total rRNA
•	rRNA_shannon: Shannon entropy score for total rRNA
•	bp_rRNA_overlap_total: Number of bases overlapping between rRNA on two strands
•	ATG, AAG, GTA, etc.: Count of respective codons in CDS.



