# GBRAP
This repository contains the codes used in the GBRAP (Genome-Based Retrieval and Analysis Parser) project. 

## Overview
A comprehensive software tool to analyse GenBank files, and an online database, housing an extensive collection of carefully curated, high-quality genome statistics generated by the software for all the organisms (latest assembly; reference or representative genome) available in the RefSeq database of NCBI. Through the database, the users can directly search or select from pre-categorized groups, the organisms of their choice and access data tables containing useful genomic information (e.g. Base counts, GC content, Shannon Entropy, Chargaff Score) separately calculated for different genomic elements (e.g. CDS, Introns, ncRNA) adding up to a collection of more than 200 genomic statistics. The data are separately displayed (if applicable) for each chromosomal, mitochondrial or chloroplast sequence. All the data can be visualised on the database itself or downloaded as CSV or Excel files. The GBRAP database is free to access without any registration and is publicly available at http://tacclab.org/gbrap/ 

## Installation and Usage
GBRAP is an open-source tool and is freely available at https://github.com/tacclab/GBRAP/blob/main/GBRAP_command_line_tool.py. It can be efficiently executed even on a standard laptop without the need for specialized hardware or configurations. 

The users only need to have the gbff file they want to analyse and the GBRAP script on their computer, and they can run the analysis through the command line. It should be noted that to maintain the quality of the data generated, the GBRAP tool is designed to only generate results for complete chromosomal and mitochondrial (and plastid or plasmid) sequences. The script will not generate any output if the input file is a draft assembly. However, if a user wishes to analyse a draft assembly, they can do so by modifying the ‘skiplocus’ filter in the Python script. 

## How to use
### 1. Download the Script: 
Download the 'GBRAP_command_line_tool.py' from this GitHub repository. Ensure Python (version 3.x) is installed on your system. All necessary libraries are included in the tool, so no additional installations are required.

### 2. Download the input file:
Download the RefSeq genome which you wish to analyse, from https://ftp.ncbi.nlm.nih.gov/genomes/refseq/ in the gbff format. The provided script is able to analyse **only chromosomal level assemblies**. Input files of draft assemblies will not generate any data. Please modify the 'skiplocus' filter if you wish to analyse a draft assembly. 

### 3. Run the analysis:
Open the command line/terminal on your computer, navigate to the folder where the script is located using the command,

```
cd /path/to/the/folder
```

Then run either of the following commands:

- If you wish to generate the output for **each chromosome** seperately (one CSV file with data of each chromosomal/mitochondrial/plasmid sequence separately in each row),
  
```
./GBRAP_command_line_tool.py -in input_file_name.gbff -out output_file_name.csv -c
```

- If you wish to generate the output for the **whole genome** at once (one CSV file with the data of the whole genome summed up to one row),
  
```
./GBRAP_command_line_tool.py -in input_file_name.gbff -out output_file_name.csv -g
```


## The Output
GBRAP returns in output a comma-separated file (CSV) that could be imported into excel. If the output file only contains the header row (no data), it means you have used a draft assembly and all the sequences were skipped due to being either 'scaffold', 'unlocalized', 'contig', 'unplaced',' patch', or 'unknown'. If there are 0s for all metrics of some genomic elements (tRNA, rRNA etc.), it means that the genome file does not contain any annotations of that element.

 The output table consists of:

•	Class: The taxonomic group (according to GBRAPs classification) of the organism.

•	Organism: The scientific name 

•	Taxon: The taxonomic classification

•	Assembly: Genome Assembly accession ID

•	Locus ID: Sequence ID

•	Version: Version number of the sequence

•	Definition: Sequence description

•	bp_chromo_A: Total number of 'A' nucleotides in the whole sequence

•	bp_chromo_T: Total number of 'T' nucleotides in the whole sequence

•	bp_chromo_C: Total number of 'C' nucleotides in the whole sequence

•	bp_chromo_G: Total number of 'G' nucleotides in the whole sequence

•	bp_chromo_N: Total number of 'N' nucleotides in the whole sequence

•	bp_chromo_tot: Total number of nucleotides in the whole sequence

•	fr_chromo_A: Frequency of 'A' nucleotides in the whole sequence 

•	fr_chromo_T: Frequency of 'T' nucleotides in the whole sequence 

•	fr_chromo_C: Frequency of 'C' nucleotides in the whole sequence 

•	fr_chromo_G: Frequency of 'G' nucleotides in the whole sequence 

•	fr_chromo_N: Frequency of 'N' nucleotides in the whole sequence 

•	GC_chromo: Percentage of ‘G’ and ‘C’ nucleotides in the whole sequence

•	topo_entropy_chromo: Topological entropy of the whole sequence

•	chargaff_pf_chromo: Chargaff's second parity rule score (method1) of the whole sequence

•	chargaff_ct_chromo: Chargaff's second parity rule score (method2) of the whole sequence

•	shannon_chromo: Shannon entropy score of the whole sequence

•	n_gene_pos: Total number of genes located on the positive strand

•	n_gene_neg: Total number of genes located on the negative strand

•	n_gene_tot: Total number of genes in the sequence

•	bp_gene_A: Total number of 'A' nucleotides in total genes

•	bp_gene_T: Total number of 'T' nucleotides in total genes

•	bp_gene_C: Total number of 'C' nucleotides in total genes

•	bp_gene_G: Total number of 'G' nucleotides in total genes

•	bp_gene_N: Total number of 'N' nucleotides in total genes

•	bp_gene_tot: Total number of nucleotides in total genes

•	fr_gene_A: Frequency of 'A' nucleotides in total genes

•	fr_gene_T: Frequency of 'T' nucleotides in total genes

•	fr_gene_C: Frequency of 'C' nucleotides in total genes

•	fr_gene_G: Frequency of 'G' nucleotides in total genes

•	fr_gene_N: Frequency of 'N' nucleotides in total genes

•	GC_gene: Percentage of ‘G’ and ‘C’ nucleotides in total genes

•	topo_entropy_gene: Topological entropy of total genes 

•	chargaff_pf_gene: Chargaff's second parity rule score (method1) of total genes

•	chargaff_ct_gene: Chargaff's second parity rule score (method2) of total genes

•	shannon_gene: Shannon entropy score of total genes

•	bp_gene_overlap_tot: Total number of nucleotides in overlapping gene regions

•	n_cds_pos: Total number of CDS located on the positive strand

•	n_cds_neg: Total number of CDS located on the negative strand

•	n_cds_tot: Total number of CDS in the sequence

•	bp_cds_A: Total number of 'A' nucleotides in total CDS (coding sequences)

•	bp_cds_T: Total number of 'T' nucleotides in total CDS

•	bp_cds_C: Total number of 'C' nucleotides in total CDS

•	bp_cds_G: Total number of 'G' nucleotides in total CDS

•	bp_cds_N: Total number of 'N' nucleotides in total CDS

•	bp_cds_tot: Total nucleotides in total CDS

•	fr_cds_A: Frequency of 'A' nucleotides in total CDS

•	fr_cds_T: Frequency of 'T' nucleotides in total CDS

•	fr_cds_C: Frequency of 'C' nucleotides in total CDS

•	fr_cds_G: Frequency of 'G' nucleotides in total CDS

•	fr_cds_N: Frequency of 'N' nucleotides in total CDS

•	GC_cds: Percentage of ‘G’ and ‘C’ nucleotides in total CDS

•	topo_entropy_cds: Topological entropy of total CDS 

•	chargaff_pf_cds: Chargaff's second parity rule score (method1) of total CDS

•	chargaff_ct_cds: Chargaff's second parity rule score (method2) of total CDS

•	shannon_cds: Shannon entropy score of total CDS

•	bp_cds_overlap_tot: Total number of nucleotides in overlapping CDS regions

•	bp_cds_intron_A: Total number of 'A' nucleotides in total CDS introns

•	bp_cds_intron_T: Total number of 'T' nucleotides in total CDS introns

•	bp_cds_intron_C: Total number of 'C' nucleotides in total CDS introns

•	'bp_cds_intron_G': Total number of 'G' nucleotides in total CDS introns

•	bp_cds_intron_N: Total number of 'N' nucleotides in total CDS introns

•	bp_cds_intron_tot: Total nucleotides in total CDS introns

•	fr_cds_intron_A: Frequency of 'A' nucleotides in total CDS introns

•	fr_cds_intron_T: Frequency of 'T' nucleotides in total CDS introns

•	'fr_cds_intron_C': Frequency of 'C' nucleotides in total CDS introns

•	fr_cds_intron_G: Frequency of 'G' nucleotides in total CDS introns

•	fr_cds_intron_N: Frequency of 'N' nucleotides in total CDS introns

•	GC_cds_intron: Percentage of ‘G’ and ‘C’ nucleotides in total CDS introns

•	topo_entropy_cds_intron: Topological entropy of total intron sequences in CDS

•	chargaff_pf_cds_intron: Chargaff's second parity rule score (method1) of total introns n CDS

•	chargaff_ct_cds_intron: Chargaff's second parity rule score (method2) of total CDS introns

•	shannon_cds_intron: Shannon entropy score of total CDS introns

•	bp_cds_intron_overlap_tot: Total number of nucleotides in overlapping CDS intron regions

•	n_ncRNA_pos: Total number of ncRNA located on the positive strand

•	n_ncRNA_neg: Total number of ncRNA located on the negative strand

•	n_ncRNA_tot: Total number of ncRNA in the sequence

•	bp_ncRNA_A: Total number of 'A' nucleotides in total ncRNA

•	 bp_ncRNA_T: Total number of 'T' nucleotides in total ncRNA

•	bp_ncRNA_C: Total number of 'C' nucleotides in total ncRNA

•	bp_ncRNA_G: Total number of 'G' nucleotides in total ncRNA

•	bp_ncRNA_N: Total number of 'N' nucleotides in total ncRNA

•	bp_ncRNA_tot: Total nucleotides in total ncRNA

•	fr_ncRNA_A: Frequency of 'A' nucleotides in total ncRNA

•	fr_ncRNA_T: Frequency of 'T' nucleotides in total ncRNA

•	fr_ncRNA_C: Frequency of 'C' nucleotides in total ncRNA

•	fr_ncRNA_G: Frequency of 'G' nucleotides in total ncRNA

•	fr_ncRNA_N: Frequency of 'N' nucleotides in total ncRNA

•	GC_ncRNA: Percentage of ‘G’ and ‘C’ nucleotides in total ncRNA

•	topo_entropy_ncRNA: Topological entropy of total ncRNA 

•	chargaff_pf_ncRNA: Chargaff's second parity rule score (method1) of total ncRNA

•	chargaff_ct_ncRNA: Chargaff's second parity rule score (method2) of total ncRNA

•	shannon_ncRNA: Shannon entropy score of total ncRNA

•	bp_ncRNA_overlap_tot: Total number of nucleotides in overlapping ncRNA regions

•	bp_nc_intron_A: Total number of 'A' nucleotides in total ncRNA introns

•	bp_nc_intron_T: Total number of 'T' nucleotides in total ncRNA introns

•	bp_nc_intron_C: Total number of 'C' nucleotides in total ncRNA introns

•	bp_nc_intron_G: Total number of 'G' nucleotides in total ncRNA introns

•	bp_nc_intron_N: Total number of 'N' nucleotides in total ncRNA introns

•	bp_nc_intron_tot: Total nucleotides in total introns between ncRNA

•	fr_nc_intron_A: Frequency of 'A' nucleotides in total ncRNA introns

•	fr_nc_intron_T: Frequency of 'T' nucleotides in total ncRNA introns

•	fr_nc_intron_C: Frequency of 'C' nucleotides in total ncRNA introns

•	fr_nc_intron_G: Frequency of 'G' nucleotides in total ncRNA introns

•	fr_nc_intron_N: Frequency of 'N' nucleotides in total ncRNA introns

•	GC_nc_intron: Percentage of ‘G’ and ‘C’ nucleotides in total ncRNA introns

•	topo_entropy_nc_intron: Topological entropy of total intron sequences in ncRNA

•	chargaff_pf_nc_intron: Chargaff's second parity rule score (method1) of total ncRNA introns

•	chargaff_ct_nc_intron: Chargaff's second parity rule score (method2) of total ncRNA introns

•	shannon_nc_intron: Shannon entropy score of total ncRNA introns

•	bp_nc_intron_overlap_tot: Total number of nucleotides in overlapping ncRNA intron regions 

•	n_tRNA_pos: Total number of tRNA located on the positive strand

•	n_tRNA_neg: Total number of tRNA located on the negative strand

•	n_tRNA_tot: Total number of tRNA in the sequence

•	bp_tRNA_A: Total number of 'A' nucleotides in total tRNA

•	bp_tRNA_T: Total number of 'T' nucleotides in total tRNA

•	bp_tRNA_C: Total number of 'C' nucleotides in total tRNA

•	bp_tRNA_G: Total number of 'G' nucleotides in total tRNA

•	bp_tRNA_N: Total number of 'N' nucleotides in total tRNA

•	bp_tRNA_tot: Total nucleotides in total tRNA

•	fr_tRNA_A: Frequency of 'A' nucleotides in total tRNA

•	fr_tRNA_T: Frequency of 'T' nucleotides in total tRNA

•	fr_tRNA_C: Frequency of 'C' nucleotides in total tRNA

•	fr_tRNA_G: Frequency of 'G' nucleotides in total tRNA

•	fr_tRNA_N: Frequency of 'N' nucleotides in total tRNA

•	GC_tRNA: Percentage of ‘G’ and ‘C’ nucleotides in total tRNA

•	topo_entropy_tRNA: Topological entropy of total tRNA 

•	chargaff_pf_tRNA: Chargaff's second parity rule score (method1) of total tRNA

•	chargaff_ct_tRNA: Chargaff's second parity rule score (method2) of total tRNA

•	shannon_tRNA: Shannon entropy score of total tRNA

•	bp_tRNA_overlap_tot: Total number of nucleotides in overlapping tRNA regions

•	n_rRNA_pos: Total number of rRNA located on the positive strand

•	n_rRNA_neg: Total number of rRNA located on the negative strand

•	n_rRNA_tot: Total number of rRNA in the sequence

•	bp_rRNA_A: Total number of 'A' nucleotides in total rRNA

•	bp_rRNA_T: Total number of 'T' nucleotides in total rRNA

•	bp_rRNA_C: Total number of 'C' nucleotides in total rRNA

•	bp_rRNA_G: Total number of 'G' nucleotides in total rRNA

•	bp_rRNA_N: Total number of 'N' nucleotides in total rRNA

•	bp_rRNA_tot: Total nucleotides in total rRNA

•	fr_rRNA_A: Frequency of 'A' nucleotides in total rRNA

•	fr_rRNA_T: Frequency of 'T' nucleotides in total rRNA

•	fr_rRNA_C: Frequency of 'C' nucleotides in total rRNA

•	fr_rRNA_G: Frequency of 'G' nucleotides in total rRNA

•	fr_rRNA_N: Frequency of 'N' nucleotides in total rRNA

•	GC_rRNA: Percentage of ‘G’ and ‘C’ nucleotides in total rRNA

•	topo_entropy_rRNA: Topological entropy of total rRNA 

•	chargaff_pf_rRNA: Chargaff's second parity rule score (method1) of total rRNA

•	chargaff_ct_rRNA: Chargaff's second parity rule score (method2) of total rRNA

•	shannon_rRNA: Shannon entropy score of total rRNA

•	bp_rRNA_overlap_tot: Total number of nucleotides in overlapping rRNA regions

•	ATG AAG GTA etc.: Count of respective codons in CDS (Codon usage)


## Chargaff’s second parity rule (PR2) calculation

In 1950, Erwin Chargaff discovered that the four nucleotides contained in a DNA double helix (A=Adenine, T=Thymine, C=Cytosine and G=Guanine) are symmetrically abundant in both strands of DNA. This symmetry was called Chargaff's first parity rule. In 1968, Chargaff also discovered that also on each DNA strand, the number of Adenines is almost equal to that of Thymines and the number of Cytosines is almost equal to that of Guanines. The first rule was easily explained by the fact that within DNA strands A matches with T, whereas C matches with G. On the single strand, however, this symmetry (Chargaff's second parity rule) is not easily explained. In 2020 four Italian researchers (Fariselli et al., 2020) discovered that this symmetry is linked to the energy of the DNA molecule which has greater stability when it has a Chargaff's second parity rule score close to one.

### † Chargaff's second parity rule score calculated using an easy method (PF)

        ABS((((#A-#T))⁄((#A+#T)))+(((#C-#G))⁄((#C+#G))))

where "#" means "number of" and A = Adenines, T = Thymines, C = Cytosines and G = Guanines. 

In this way the perfect Chargaff’s second parity rule score is zero such as in case of “ATGC”. The minimum value depends on the length of the sequence. 
Python 3 code is as follows:
```
        def chargaff_pf(self,sequence):
                '''calculates chargaff score PF'''
                counts = Counter(sequence)
                
                def safe_pf(x,y):
                    if x==0 and y==0:
                        return 0
                    return abs((x-y)/(x+y))
                
                a_t=safe_pf(counts['A'], counts['T'])
                c_g=safe_pf(counts['C'], counts['G'])
                
                PF=a_t+c_g
                
                return PF
```

### †  Chargaff's second parity rule score calculated using Cristian Taccioli (CT) method 

        (((#A)/(#T)+ (#C)/(#G)))/2,where #T and #G ≠0
        
The bases with the highest value must be placed in the denominator.
 In this example A and C have a lower value than T and G respectively. 

"#" means "number of" and A = Adenines, T = Thymines, C = Cytosines and G = Guanines.
 Using this equation Chargaff’s second parity rule score is always between zero and one, where one is the maximum value. For example, the sequence “ATGC” has a perfect score which is one because the number of A is equal to the number of T, whereas the number of C is equal to the number of G. This score does not depend on sequence length. Chargaff’s second parity rule calculated using C.T method can be calculated in Python 3 as follow:

```
         def chargaff_ct(self,sequence):
                ''' calculates chargaff score CT '''
                counts = Counter(sequence)
        
                def safe_ratio(x, y):
                    if x == 0 or y == 0:
                        return 0
                    return min(x, y) / max(x, y)
        
                a_t_ratio = safe_ratio(counts['A'], counts['T'])
                c_g_ratio = safe_ratio(counts['C'], counts['G'])
        
                CT = (a_t_ratio + c_g_ratio) / 2
```

## Entropy scores calculation
The concept of entropy was introduced in the early 19th century by Rudolf Julius Emanuel Clausius. It represents a characteristic quantity of the state of a physical system capable of expressing the ability of the system itself to be able to proceed to spontaneous transformations and, consequently, the loss of ability to do work when such transformations occur. In simplified terms, the value of entropy increases when the system undergoes spontaneous variations and therefore loses part of its ability to undergo such variations and perform work. In 1872 Ludwig Boltzmann generalized this concept through the study of statistical mechanics by defining entropy as the degree of disorder of a system. In 1948 Claude Elwood Shannon equated the degree of inaccuracy of a message with disorder. For Shannon, in fact, the entropy of information was the degree of complexity of a message that represents the minimum average number of symbols necessary for the encoding of the message itself.

### †  Shannon score 

The term entropy in information sciences was introduced by Shannon in the paper "A Mathematical Theory of Communication" (Shannon, 1948). Shannon entropy is calculated as follows:

      H(X) = - ∑ⁿᵢ₌₁ P(xᵢ) log P(xᵢ)
  
where P is the frequency of nucleotides

Python 3 code for Shannon entropy is:
```
        def shannon (self,seq,base_counts):
                ''' calculates shanon entropy'''
                nt=base_counts
                if nt[0]==0:
                    return ''
                else:
                    pA = float(nt[1]/nt[0])
                    pC = float(nt[2]/nt[0])
                    pG = float(nt[3]/nt[0])
                    pT = float(nt[4]/nt[0])
                    
                    if pA == 0:
                        pA=1
                    if pC == 0:
                        pC=1
                    if pG == 0:
                        pG=1
                    if pT == 0:
                        pT=1
                    
                    return -((pA*math.log2(pA))+(pT*math.log2(pT))+(pC*math.log2(pC))+(pG*math.log2(pG)))
```
          

### † Topological entropy score

Topological entropy is a nonnegative real number that is capable of measuring the complexity of a message. Topological entropy was first introduced in 1965 by Adler, Konheim and McAndrew (Adler et al. 1965). In 2011 Koslicki has defined a new approximation to topological entropy free from the finite sample effects and high dimensionality problems. The formula and code can be retrieved from Koslicki, 2011 (Koslicki et al. 2011).

```
        def topology (self,seq): 
                '''calculates topological entropy'''
                s= seq.replace("N","") ##Removes the N in the sequence
                n = len(set(s)) # number of bases, this is usually equal to 4
                length = len(s)
                if len(s) != 0 and n > 1: 
                    logg = math.floor(math.log(length,n))
                    neww = (s[:(n**logg + logg)]).lower()
                    result_mr = math.log(len(set([neww[i:(i+logg)] for i in range(1, n**logg+1)])),n)/logg #the topological entropy score
                else:
                    result_mr = ''
                return result_mr
```

## Reference
1. Adler RL, Konheim AG, McAndrew MH. 1965. Topological entropy. Trans Am Math Soc 114:309–319.

2. Fariselli P, Taccioli C, Pagani L, Maritan A. 2020. DNA sequence symmetries from randomness: the origin of the Chargaff second parity rule. Brief. Bioinform.:1–10.

3. Goldfarb T, K odali VK, ant Pujar S, ac hesla Bro er V, Robber tse B, F ar rell CM, Oh D-H, Astashyn A, Er molaev O, Haddad D, et al. 2025. NCBI RefSeq: reference sequence standards through 25 years of curation and annotation. Nucleic Acids Res 53:D243–D257.

4. Koslicki D. 2011. Topological entropy of DNA sequences. Bioinformatics 27:1061–1067.

5. Shannon CE. 1948. A Mathematical Theory of Communication. The Bell System Technical Journal 27:623–656.


# DISCLAIMER

We will have no responsibility or liability in relation to any loss or damage that you incur, including damage to your software or hardware, arising from your use of GBRAP (github/tacclab/GBRAP).


